packages/salesforcedx-vscode-apex/src/oas/esrHandler.ts

```diff
@@ -0,0 +1,447 @@
+    promptGenerationOrchestrator: PromptGenerationOrchestrator
+  ) {
+    const orgVersion = await (await WorkspaceContextUtil.getInstance().getConnection()).retrieveMaxApiVersion();
+    let existingContent;
+    let namedCredential;
+    if (fs.existsSync(this.newPath)) {
+      existingContent = fs.readFileSync(this.newPath, 'utf8');
+    }
+
+    //Step 1: Let user choose or enter a named credential
+    if (!this.isVersionGte(orgVersion, '63.0')) namedCredential = await this.showNamedCredentialsQuickPick();
+
+    //Step 2: Build the content of the ESR Xml file
+    const updatedContent = await this.buildESRXml(existingContent, namedCredential);
+
+    //Step 3: Write OpenAPI Document to File
+    try {
+      fs.writeFileSync(this.newPath, updatedContent);
+      await vscode.workspace.openTextDocument(this.newPath).then((newDocument: vscode.TextDocument) => {
+        void vscode.window.showTextDocument(newDocument);
+      });
+      if (this.isESRDecomposed) {
+        await vscode.workspace
+          .openTextDocument(this.replaceXmlToYaml(this.newPath))
+          .then((newDocument: vscode.TextDocument) => {
+            void vscode.window.showTextDocument(newDocument);
+          });
+      }
+    } catch (error) {
+      throw new Error(nls.localize('artifact_failed', error.message));
+    }
+
+    //Step 2: Gather metrics
+    const props = {
+      isClass: `${isClass}`,
+      overwrite: `${this.overwrite}`
+    };
+
+    const [errors, warnings, infos, hints, total] = summarizeDiagnostics(this.processedOasResult.errors);
+
+    const measures = {
+      generationDuration: (await getTelemetryService()).hrTimeToMilliseconds(generationHrDuration),
+      biddedCallCount: promptGenerationOrchestrator.strategy?.biddedCallCount,
+      llmCallCount: promptGenerationOrchestrator.strategy?.llmCallCount,
+      generationSize: promptGenerationOrchestrator.strategy?.maxBudget,
+      documentTtlProblems: total,
+      documentErrors: errors,
+      documentWarnings: warnings,
+      documentInfo: infos,
+      documentHints: hints
+    };
+
+    // Step 3: If the user chose to merge, open a diff between the original and new ESR files
+    this.openDiffFiles();
+
+    // Step: 4 Create entries in problems tab for generated file
+    createProblemTabEntriesForOasDocument(
+      this.isESRDecomposed ? this.replaceXmlToYaml(this.newPath) : this.newPath,
+      this.processedOasResult,
+      this.isESRDecomposed
+    );
+    return { props, measures };
+  }
+
+  /**
+   * Prompts the user to select or enter a named credential.
+   * This method queries the available named credentials and presents them in a quick pick menu.
+   * If the user selects "Enter new named credential", they are prompted to enter the name manually.
+   *
+   * @returns A promise that resolves to the selected or entered named credential, or undefined if none was selected.
+   */
+  private async showNamedCredentialsQuickPick(): Promise<string | undefined> {
```

Comment:
@CristiCanizales this function is going to be hard to unit test, given its dependence on connections and vscode runtime for UI interaction. please consider relocating those pieces to separate functions that will be easier to mock